js_modules/dagit/packages/core/src/nav/useCodeLocationsStatus.tsx

```diff
@@ -85,37 +86,38 @@ export const useCodeLocationsStatus = (skip = false): StatusAndMessage | null =>
     }
   }, [onClickViewButton, refetch]);
 
-  // Given the previous and current code locations, determine whether to show a) a loading spinner
-  // and/or b) a toast indicating that a code location is being reloaded.
-  React.useEffect(() => {
-    const previousEntries =
-      previousData?.locationStatusesOrError?.__typename === 'WorkspaceLocationStatusEntries'
-        ? previousData?.locationStatusesOrError.entries
-        : [];
-    const currentEntries =
+  const onLocationUpdate = (data: CodeLocationStatusQuery) => {
+    // Given the previous and current code locations, determine whether to show a) a loading spinner
+    // and/or b) a toast indicating that a code location is being reloaded.
+    const entries =
       data?.locationStatusesOrError?.__typename === 'WorkspaceLocationStatusEntries'
         ? data?.locationStatusesOrError.entries
         : [];
 
-    const _build_entries_by_name = (
-      entries: LocationStatusEntry[],
-    ): {[key: string]: LocationStatusEntry} => {
-      const entries_by_name = {};
-      entries.forEach((entry) => {
-        entries_by_name[entry.id] = entry;
-      });
-      return entries_by_name;
-    };
-    const previousEntriesByName = _build_entries_by_name(previousEntries);
-    const currentEntriesByName = _build_entries_by_name(currentEntries);
-    const hasUpdatedEntries =
-      currentEntries.length &&
-      currentEntries.some(
-        (entry) =>
-          !(entry.id in previousEntriesByName) ||
-          previousEntriesByName[entry.id].updateTimestamp <
-            currentEntriesByName[entry.id].updateTimestamp,
-      );
+    let hasUpdatedEntries = entries.length !== Object.keys(state.entriesById || {}).length;
+    const entriesById: {[key: string]: LocationStatusEntry} = {};
+    entries.forEach((entry) => {
+      const previousEntry = state.entriesById && state.entriesById[entry.id];
+      const entryIsUpdated =
+        !previousEntry ||
+        previousEntry.updateTimestamp < entry.updateTimestamp ||
+        previousEntry.loadStatus !== entry.loadStatus;
+      hasUpdatedEntries = hasUpdatedEntries || entryIsUpdated;
+      entriesById[entry.id] = entryIsUpdated
+        ? {
+            id: entry.id,
+            loadStatus: entry.loadStatus,
+            name: entry.name,
+            updateTimestamp: entry.updateTimestamp,
+          }
+        : previousEntry;
+    });
+    const previousEntriesById = state.entriesById;
+    const currentEntries = Object.values(entriesById || {});
+    const previousEntries = Object.values(previousEntriesById || {});
+    if (hasUpdatedEntries) {
+      setState({entriesById});
+    }
```

Comment:
This could probably be extracted out into a function (accepting the old and new entries as params) that you could write some tests against, with the `entriesById` being returned to be used in the setter.